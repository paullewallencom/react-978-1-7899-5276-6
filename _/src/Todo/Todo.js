import React, { Component } from "react";
import { Button, ButtonGroup } from "reactstrap";
import styles from "./Todo.module.scss";

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      done: props.done,
      critical: props.critical
    };

    this.markAsDone = this.markAsDone.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
    this.markCritical = this.markCritical.bind(this);
  }
  markCritical() {
    this.setState({ critical: true });
  }
  markAsDone() {
    this.setState({ done: true });
  }
  removeTodo() {
    this.props.removeTodo(this.props.id);
  }
  cssClasses() {
    let classes = [];
    if (this.state.critical) {
      classes = [styles.critical];
    } else {
      classes = [styles.todo];
    }
    if (this.state.done) {
      classes = [...classes, styles.done];
    }
    return classes.join(" ");
  }
  render() {
    return (
      <div className={this.cssClasses()}>
        {this.props.description}
        <br />
        <hr className={styles.hr} />
        <ButtonGroup>
          <Button
            className="MarkDone"
            onClick={this.markAsDone}
            color="success"
          >
            Mark as Done
          </Button>
          <Button
            className="RemoveTodo"
            onClick={this.removeTodo}
            color="warning"
          >
            Remove Me
          </Button>
          <Button
            className="MarkCritical"
            onClick={this.markCritical}
            color="danger"
          >
            Mark as Critical
          </Button>
        </ButtonGroup>
      </div>
    );
  }
}

export default Todo;
