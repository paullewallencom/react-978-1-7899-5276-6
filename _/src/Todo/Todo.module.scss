@import "../shared";

.todo {
  border: 2px solid black;
  text-align: center;
  background: $todo-normal;
  color: #333;
  margin: 20px;
  padding: 20px;
}

.done {
  background: $todo-complete;
}

.hr {
  border: 2px solid red;
}

.critical {
  composes: todo;
  background: $todo-critical;
}
