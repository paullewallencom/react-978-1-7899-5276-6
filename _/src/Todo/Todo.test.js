import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import Todo from "./Todo";

describe(Todo, () => {
  const description = "New Todo";
  const mockRemoveTodo = jest.fn();
  const component = shallow(
    <Todo
      description={description}
      removeTodo={mockRemoveTodo}
      critical={false}
      done={false}
      id={1}
    />
  );

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Todo />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders and matches our snapshot", () => {
    const component = renderer.create(<Todo description="Yo" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a Todo component", () => {
    expect(component.contains(<div className="Todo" />));
  });

  it("contains the description", () => {
    expect(component.text()).toContain(description);
  });

  it("marks the Todo as done", () => {
    component.find("Button.MarkDone").simulate("click");
    expect(component.state("done")).toEqual(true);
  });

  it("calls the mock remove function", () => {
    component.find("Button.RemoveTodo").simulate("click");
    expect(mockRemoveTodo).toHaveBeenCalled();
  });

  it("marks the Todo as critical", () => {
    expect(component.state("critical")).toEqual(false);
    component.find("Button.MarkCritical").simulate("click");
    expect(component.state("critical")).toEqual(true);
  });
});
