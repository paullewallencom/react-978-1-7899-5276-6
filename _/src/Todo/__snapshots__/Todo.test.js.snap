// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo renders and matches our snapshot 1`] = `
<div
  className="todo"
>
  Yo
  <br />
  <hr
    className="hr"
  />
  <div
    className="btn-group"
    role="group"
  >
    <button
      aria-label={null}
      className="MarkDone btn btn-success"
      onClick={[Function]}
      type="button"
    >
      Mark as Done
    </button>
    <button
      aria-label={null}
      className="RemoveTodo btn btn-warning"
      onClick={[Function]}
      type="button"
    >
      Remove Me
    </button>
    <button
      aria-label={null}
      className="MarkCritical btn btn-danger"
      onClick={[Function]}
      type="button"
    >
      Mark as Critical
    </button>
  </div>
</div>
`;
